### PR TITLE
USD base currency

### PR DESCRIPTION
**Pull request by mayankamencherla**
__Need to test the following__
1. USD payment with INR base currency
- [x] Order creation is correct
- [x] checkout loads the correct amounts and display amounts in each case
2. INR payment normal flow
- [x] Order creation is correct
- [x] checkout loads the correct amounts and display amounts in each case
3. USD payment with USD base currency
- [x] Order creation is correct
- [x] checkout loads the correct amounts and display amounts in each case

### CODE REVIEW

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 1]
```diff
@@ -72,16 +72,18 @@ protected function verifyOrderAmount($razorpayOrderId, $order)
...
     protected function getExpectedRazorpayOrderData($order)
     {
-        $amount  = (int) ($order->getBaseGrandTotal() * 100);
+        $orderCurrency = $order->getBaseCurrencyCode();
 
-        $orderId = $order->getRealOrderId();
+        $orderAmount = (int) ($order->getBaseGrandTotal() * 100);
 
-        $currency = Razorpay_Payments_Model_Paymentmethod::CURRENCY;
+        if ($orderCurrency !== Razorpay_Payments_Model_Paymentmethod::CURRENCY) {
```
Comment (ccrims0n) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: @mayankamencherla change constant to `BASE_CURRENCY`?

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 2]
```diff
@@ -72,16 +72,18 @@ protected function verifyOrderAmount($razorpayOrderId, $order)
...
+            $orderAmount = $this->getOrderAmountInInr($orderAmount, $orderCurrency);
+        }
 
         $data = array(
-            'receipt'  => $orderId,
-            'amount'   => $amount,
-            'currency' => $currency,
+            'receipt'  => $order->getRealOrderId(),
+            'amount'   => $orderAmount,
+            'currency' => 'INR',
```
Comment (ccrims0n) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: Use constant here.

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 3]
```diff
@@ -100,11 +102,20 @@ protected function getRazorpayOrderData($order)
         return $data;
     }
 
+    protected function getOrderAmountInInr($orderAmount, $orderCurrency)
+    {
+        $url = "http://api.fixer.io/latest?base=$orderCurrency";
+
+        $rates = json_decode(file_get_contents($url), true);
+
+        return $orderAmount * $rates['rates']['INR'];
```
Comment (ccrims0n) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: @mayankamencherla this will return a float.

File: app/code/community/Razorpay/Payments/Helper/Data.php (hunk 1 above)
Comment (mayankamencherla) [reply app/code/community/Razorpay/Payments/Helper/Data.php]: @ccrims0n We now support USD as base currency as well. So INR isn't the only base currency.

File: app/code/community/Razorpay/Payments/Helper/Data.php (hunk 3 above)
Comment (mayankamencherla) [reply app/code/community/Razorpay/Payments/Helper/Data.php]: @ccrims0n Using 
```
return (int) round($orderAmount * $rates['rates']['INR'], 2);
```

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 4]
```diff
@@ -100,11 +103,20 @@ protected function getRazorpayOrderData($order)
         return $data;
     }
 
+    protected function getOrderAmountInInr($orderAmount, $orderCurrency)
+    {
+        $url = "http://api.fixer.io/latest?base=$orderCurrency";
+
+        $rates = json_decode(file_get_contents($url), true);
+
+        return (int) round($orderAmount * $rates['rates']['INR'], 2);
```
Comment (ccrims0n) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: Why are we using `round` and also typecasting it to the integer.

File: js/razorpay/razorpay-utils.js [hunk 5]
```diff
@@ -5,47 +5,69 @@ function RazorpayUtils(options) {
...
+
+                var url = 'http://api.fixer.io/latest?base=' + this.options.base_currency;
 
-        checkout = new Razorpay(options);
+                // Setting amount based on current currency rate
+                var response = this.httpGetAsync(url);
 
-        checkout.open();
-    }
+                options['amount'] = JSON.parse(response).rates.INR * options['amount'];
```
Comment (ccrims0n) [inline js/razorpay/razorpay-utils.js]: @mayankamencherla Why are we doing conversion here?

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 6]
```diff
@@ -100,13 +103,22 @@ protected function getRazorpayOrderData($order)
...
     {
-        $amount             = (int) ($order->getBaseGrandTotal() * 100);
-        $base_currency      = $order->getBaseCurrencyCode();
-        $quote_currency     = $order->getCurrencyCode();
-        $quote_amount       = round($order->getGrandTotal(), 2);
+        $url = "http://api.fixer.io/latest?base=$orderCurrency";
+
+        $rates = json_decode(file_get_contents($url), true);
+
+        return (int) ($orderAmount * $rates['rates']['INR']);
```
Comment (ccrims0n) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: we should use `ceil()` here

**Review by ccrims0n — CHANGES_REQUESTED** [review]
We shouldn't calculate the order price again.

**Comment by mayankamencherla** [comment]
@ccrims0n Setting and getting order amount in the session.

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 7]
```diff
@@ -100,13 +103,31 @@ protected function getRazorpayOrderData($order)
...
+     */
+    protected function getOrderAmountInInr($orderAmount, $orderCurrency)
     {
-        $amount             = (int) ($order->getBaseGrandTotal() * 100);
-        $base_currency      = $order->getBaseCurrencyCode();
-        $quote_currency     = $order->getCurrencyCode();
-        $quote_amount       = round($order->getGrandTotal(), 2);
+        $amount = Mage::getSingleton('core/session')->getOrderAmount();
 
+        if ($amount === null)
```
Comment (ccrims0n) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: @mayankamencherla this check shouldn't be here.

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 8]
```diff
@@ -72,16 +72,19 @@ protected function verifyOrderAmount($razorpayOrderId, $order)
...
-        $amount  = (int) ($order->getBaseGrandTotal() * 100);
+        $orderCurrency = $order->getBaseCurrencyCode();
 
-        $orderId = $order->getRealOrderId();
+        $orderAmount = (int) ($order->getBaseGrandTotal() * 100);
 
-        $currency = Razorpay_Payments_Model_Paymentmethod::CURRENCY;
+        if ($orderCurrency !== Razorpay_Payments_Model_Paymentmethod::CURRENCY)
+        {
+            $orderAmount = $this->getOrderAmountInInr($orderAmount, $orderCurrency);
```
Comment (ccrims0n) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: @mayankamencherla fetch from session here.

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 9]
```diff
@@ -100,13 +103,36 @@ protected function getRazorpayOrderData($order)
...
+        // For create step, we always re-calculate the INR amount
+        // For validation step, we only calculate if the amount is not stored in the session
+        //
+        if ($create === true)
+        {
+            $url = "http://api.fixer.io/latest?base=$orderCurrency";
+
+            $rates = json_decode(file_get_contents($url), true);
+
+            $amount = ceil($orderAmount * $rates['rates'][Razorpay_Payments_Model_Paymentmethod::CURRENCY]);
```
Comment (ccrims0n) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: please add an explicit typecast.

File: js/razorpay/razorpay-utils.js [hunk 10]
```diff
@@ -5,47 +5,47 @@ function RazorpayUtils(options) {
...
 
-    this.placeOrder = function(onSuccess, onDismiss, formId, paymentIdField)
-    {
-        if (!formId) {
-            formId = "razorpay";
-        }
-        if (!paymentIdField) {
-            paymentIdField = "razorpay_payment_id";
-        }
+        this.placeOrder = function(onSuccess, onDismiss, formId, paymentIdField) {
```
Comment (ccrims0n) [inline js/razorpay/razorpay-utils.js]: @mayankamencherla why is this change needed?

Comment (mayankamencherla) [reply js/razorpay/razorpay-utils.js]: @ccrims0n JS convention to have opening brace on the same line.

Comment (ccrims0n) [reply js/razorpay/razorpay-utils.js]: @mayankamencherla please move this to a separate pr.

**Review by ccrims0n — APPROVED** [review]
please make the changes, test and merge
